lib/tbot/service_outputs.go

```diff
@@ -49,29 +49,143 @@ import (
+	// server with similar requests
+	drc := &outputRenewalCache{
+		client: client,
+		cfg:    s.cfg,
+	}
+
+	// Determine the default role list based on the bot role. The role's
+	// name should match the certificate's Key ID (user and role names
+	// should all match bot-$name)
+	botResourceName := botIdentity.X509Cert.Subject.CommonName
+	defaultRoles, err := fetchDefaultRoles(ctx, client, botResourceName)
+	if err != nil {
+		s.log.WithError(err).Warnf("Unable to determine default roles, no roles will be requested if unspecified")
+		defaultRoles = []string{}
+	}
+
+	// Next, generate impersonated certs
+	for _, output := range s.cfg.Outputs {
+		s.log.WithFields(logrus.Fields{
+			"output": output,
+		}).Info("Generating output.")
+
+		dest := output.GetDestination()
+		// Check the ACLs. We can't fix them, but we can warn if they're
+		// misconfigured. We'll need to precompute a list of keys to check.
+		// Note: This may only log a warning, depending on configuration.
+		if err := dest.Verify(identity.ListKeys(identity.DestinationKinds()...)); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Ensure this destination is also writable. This is a hard fail if
+		// ACLs are misconfigured, regardless of configuration.
+		// TODO: consider not making these a hard error? e.g. write other
+		// destinations even if this one is broken?
+		if err := identity.VerifyWrite(ctx, dest); err != nil {
+			return trace.Wrap(err, "testing output destination: %s", output)
+		}
+
+		impersonatedIdentity, impersonatedClient, err := s.generateImpersonatedIdentity(
+			ctx, client, botIdentity, output, defaultRoles,
+		)
+		if err != nil {
+			return trace.Wrap(err, "generating impersonated certs for output: %s", output)
+		}
+		defer impersonatedClient.Close()
+
+		s.log.WithFields(logrus.Fields{
+			"identity": describeTLSIdentity(s.log, impersonatedIdentity),
+			"output":   output,
+		}).Debug("Fetched identity for output.")
+
+		// Create a destination provider to bundle up all the dependencies that
+		// a destination template might need to render.
+		dp := &outputProvider{
+			outputRenewalCache: drc,
+			impersonatedClient: impersonatedClient,
+		}
+
+		if err := output.Render(ctx, dp, impersonatedIdentity); err != nil {
+			s.log.WithError(err).Warnf("Failed to render output %s", output)
+			return trace.Wrap(err, "rendering output: %s", output)
+		}
+
+		s.log.WithFields(logrus.Fields{
+			"output": output,
+		}).Info("Generated output.")
+	}
+
+	return nil
+}
+
+func (s *outputsService) Run(ctx context.Context) error {
+	ctx, span := tracer.Start(ctx, "outputsService/Run")
+	defer span.End()
+
+	reloadCh, unsubscribe := s.reloadBroadcaster.subscribe()
+	defer unsubscribe()
+
+	s.log.Infof(
 		"Beginning output renewal loop: ttl=%s interval=%s",
-		b.cfg.CertificateTTL,
-		b.cfg.RenewalInterval,
+		s.cfg.CertificateTTL,
+		s.cfg.RenewalInterval,
 	)
 
-	ticker := time.NewTicker(b.cfg.RenewalInterval)
+	ticker := time.NewTicker(s.cfg.RenewalInterval)
```

Comment:
Bit of an aside that doesn't need to be solved here:

I know this isn't a new logic change and the main downsides were mitigated by #24228, but it could be neat to be able to subscribe to cert update events from the bot identity service to avoid needing a totally separate renewal loop for outputs.

Reply:
Agreed